README.md

```diff
@@ -1,4 +1,134 @@
+
+We'll discuss each steps in detail.
+
+### Prepare the Docker image
+
+1. Try to download a Docker image specified by `TRAMPOLINE_IMAGE`
+   environment variable.
+2. If `TRAMPOLINE_DOCKERFILE` environment variable is specified, it
+   will build the Docker image from that Dockerfile. If there's a
+   downloaded Docker image, Trampoline V2 uses the downloaded image as
+   a cache.
+
+We recomnend you have the Dockerfile in the repo and specify
+`TRAMPOLINE_DOCKERFILE`. This will allow you to create a single pull
+request containing changes in tests as well as changes in the
+Dockerfile.
+
+### Run Docker with appropriate flags to run the test
+
+* Mounting the source at /workspace
+* Add appropriate environment variables
+* Use invoker's uid and the Docker gid on the host.
+* Runs a command specified by `TRAMPOLINE_BUILD_FILE` environment
+  variable.
+* Trampoline V2 will exit with the same exit code as the build file.
+
+#### Environment variables which will be passed down into the container
+
+The following environment variables are passed down into the container.
+
+* `RUNNING_IN_CI`
+  Tells scripts whether they are running as part of CI or not.
+* `TRAMPOLINE_CI`
+  Indicates which CI system we're in.
+* `TRAMPOLINE_V2`
+  (true|false) Indicates we're running trampoline_v2.
+* `KOKORO_BUILD_NUMBER`
+* `KOKORO_BUILD_ID`
+* `KOKORO_JOB_NAME`
+* `KOKORO_GIT_COMMIT`
+* `KOKORO_GITHUB_COMMIT`
+* `KOKORO_GITHUB_PULL_REQUEST_NUMBER`
+* `KOKORO_GITHUB_PULL_REQUEST_COMMIT`
+* `KOKORO_GITHUB_COMMIT_URL`
+* `KOKORO_GITHUB_PULL_REQUEST_URL`
+
+### Upload the newly built Docker image
+
+If `TRAMPOLINE_IMAGE_UPLOAD` environment variable is set to `true`,
+Trampoline V2 will try to upload the newly built image to
+`TRAMPOLINE_IMAGE` after the successful build.
+
+This optional step allows you to upload the built image only after a
+full build succeeds. To make this step work, you need to set up
+authentication.
+
+### Authentication
+
+To set up authentication on Kokoro (Google internal CI system), we
+recommend that you attach a service account to your VM instance and
+add appropriate permissions.
+
+On other CI systems, you need to securely pass a service account json
+file and set `TRAMPOLINE_SERVICE_ACCOUNT` environment variable
+pointing to that file.
+
+In either case, you need at least read permission on the Google
+Container Registry artifact bucket. You also need write permission on
+the bucket if you use the uploading feature.
+
+### Environment variables
+
+By design, all the parameters to Trampoline V2 are passed by
+environment variables.
+
+There are two required environment variables:
+* `TRAMPOLINE_IMAGE`: The docker image to use. You can use Docker
+  image URL on anywhere, but if you also want to upload the docker
+  image, use an URL on Google Container Registry
+  (gcr.io/PROJ/IMAGE_NAME).
+* `TRAMPOLINE_BUILD_FILE`: The script to run in the docker
+  container. It should be a relative path from the git root.
```

Comment:
Similar to above, both are not required - you need one of them set?

Reply:
I guess you're talking about `TRAMPOLINE_DOCKERFILE`?

These envvars are mandatory now.

Reply:
Oops, yeah misread.